php/data/MapIterator.php

```diff
@@ -0,0 +1,30 @@
+<?php
+/**
+ * This file defines the concept of a data Provisioner.
+ *
+ * PHP Version 7
+ *
+ * @category   Data
+ * @package    Main
+ * @subpackage Data
+ * @author     Dave MacFarlane <david.macfarlane2@mcgill.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link       https://www.github.com/aces/Loris/
+ */
+namespace LORIS\Data;
+
+class MapIterator extends \IteratorIterator {
```

Comment:
For the sake of discussion, it should be noted that `IteratorIterator` will mutate itself when `next()` is called. Unfortunately you cannot go against the php spec (return type `void`). It's just something to keep in mind since we're trying to have immutable data.